src/django_otp/plugins/otp_email/models.py

```diff
@@ -57,34 +63,50 @@ def generate_challenge(self, extra_context=None):
         :type extra_context: dict
 
         """
-        self.generate_token(valid_secs=settings.OTP_EMAIL_TOKEN_VALIDITY)
-
-        context = {'token': self.token, **(extra_context or {})}
-        if settings.OTP_EMAIL_BODY_TEMPLATE:
-            body = Template(settings.OTP_EMAIL_BODY_TEMPLATE).render(Context(context))
-        else:
-            body = get_template(settings.OTP_EMAIL_BODY_TEMPLATE_PATH).render(context)
-
-        if settings.OTP_EMAIL_BODY_HTML_TEMPLATE:
-            body_html = Template(settings.OTP_EMAIL_BODY_HTML_TEMPLATE).render(
-                Context(context)
-            )
-        elif settings.OTP_EMAIL_BODY_HTML_TEMPLATE_PATH:
-            body_html = get_template(settings.OTP_EMAIL_BODY_HTML_TEMPLATE_PATH).render(
-                context
+        generate_allowed, data_dict = self.generate_is_allowed()
+        if generate_allowed:
+            self.generate_token(valid_secs=settings.OTP_EMAIL_TOKEN_VALIDITY)
```

Comment:
This method is now doing two big things. Perhaps it's time to move the body of `if generate_allowed:` into a private method for clarity. See https://github.com/django-otp/django-otp-twilio/blob/9e1aab519c926a3d281292d549ce2040c821fe1a/src/otp_twilio/models.py#L55, e.g.

Reply:
Created a private method for that part ([dbe61d7](https://github.com/django-otp/django-otp/pull/122/commits/dbe61d7fe664ad0fc7ebeb74b630ea682cc970b6)).